txhelpers/txhelpers.go

```diff
@@ -249,6 +249,191 @@ func TxInvolvesAddress(msgTx *wire.MsgTx, addr string, c VerboseTransactionGette
+				addrs[addr] = false // not new to the address store
+			}
+			addrOuts.Outpoints = append(addrOuts.Outpoints, op)
+		}
+	}
+	return
+}
+
+// TxPrevOutsByAddr sets the PrevOuts field for the AddressOutpoints stored in
+// the MempoolAddressStore. For addresses not yet present in the
+// MempoolAddressStore, a new AddressOutpoints is added to the store. The
+// provided MempoolAddressStore must be initialized. A VerboseTransactionGetter
+// is required to retrieve the pkScripts for the previous outpoints. The number
+// of consumed previous outpoints that paid addresses in the provided
+// transaction are counted and returned. The addresses in the previous outpoints
+// are listed in the output addrs map, where the value of the stored bool
+// indicates the address is new to the MempoolAddressStore.
+func TxPrevOutsByAddr(txAddrOuts MempoolAddressStore, txnsStore TxnsStore, msgTx *wire.MsgTx, c VerboseTransactionGetter, params *chaincfg.Params) (newPrevOuts int, addrs map[string]bool) {
+	if txAddrOuts == nil {
+		panic("TxPrevOutAddresses: input map must be initialized: map[string]*AddressOutpoints")
+	}
+	if txnsStore == nil {
+		panic("TxPrevOutAddresses: input map must be initialized: map[string]*AddressOutpoints")
+	}
+
+	// Send all the raw transaction requests
+	type promiseGetRawTransaction struct {
+		result rpcclient.FutureGetRawTransactionVerboseResult
+		inIdx  int
+	}
+	promisesGetRawTransaction := make([]promiseGetRawTransaction, 0, len(msgTx.TxIn))
+
+	for inIdx, txIn := range msgTx.TxIn {
+		hash := &txIn.PreviousOutPoint.Hash
+		if zeroHash.IsEqual(hash) {
+			continue // coinbase or stakebase
+		}
+		promisesGetRawTransaction = append(promisesGetRawTransaction, promiseGetRawTransaction{
+			result: c.GetRawTransactionVerboseAsync(hash),
+			inIdx:  inIdx,
+		})
+	}
+
+	addrs = make(map[string]bool)
+
+	// For each TxIn of this transaction, inspect the previous outpoint.
+	for i := range promisesGetRawTransaction {
+		// Previous outpoint for this TxIn
+		inIdx := promisesGetRawTransaction[i].inIdx
+		prevOut := &msgTx.TxIn[inIdx].PreviousOutPoint
+		hash := prevOut.Hash
+
+		prevTxRaw, err := promisesGetRawTransaction[i].result.Receive()
+		if err != nil {
+			fmt.Printf("Unable to get raw transaction for %v: %v\n", hash, err)
+			return
+		}
+
+		if prevTxRaw.Txid != hash.String() {
+			fmt.Printf("TxPrevOutsByAddr error: %v != %v", prevTxRaw.Txid, hash.String())
+			continue
+		}
+
+		prevTx, err := MsgTxFromHex(prevTxRaw.Hex)
+		if err != nil {
+			fmt.Printf("TxPrevOutsByAddr: MsgTxFromHex failed: %s\n", err)
+			continue
+		}
+
+		// prevOut.Index indicates which output.
+		txOut := prevTx.TxOut[prevOut.Index]
+		// Extract the addresses from this output's PkScript.
+		_, txAddrs, _, err := txscript.ExtractPkScriptAddrs(
+			txOut.Version, txOut.PkScript, params)
+		if err != nil {
+			fmt.Printf("TxPrevOutsByAddr: ExtractPkScriptAddrs: %v\n", err.Error())
+			continue
+		}
+
+		if len(txAddrs) == 0 {
+			continue
```

Comment:
For a previous outpoint, perhaps not.  Clearly and output can be something like an OP_RETURN, but here these are prevouts that are consumed, they will usually have an address decode from the PkScript.  In the case of a coinbase or stakebase input, there is no PkScript so it shouldn't get to this point because of this check: https://github.com/decred/dcrdata/pull/1222/files/ba6324d6b0ae5f469b59ffac951e45511f59e54b#diff-d28ceeced3a3ca512173e7b6df946805R338.  How about I put a warning here before the `continue` in case there is something we haven't though of?